File: probe-go/circular_buffer.go
package main

import (
	"sync/atomic"
	"unsafe"
)

// ring buffer, pre-allocated at startup
// atomic ops keep it thread-safe without locks
type CircularBuffer struct {
	buffer     []float64
	capacity   int64
	writeIndex atomic.Int64
	size       atomic.Int64
}

func NewCircularBuffer(capacity int64) *CircularBuffer {
	return &CircularBuffer{
		buffer:   make([]float64, capacity),
		capacity: capacity,
	}
}

// overwrite oldest when full
func (cb *CircularBuffer) Write(latencyMs float64) {
	idx := cb.writeIndex.Add(1) - 1
	// unsafe write to skip allocation
	*(*float64)(unsafe.Pointer(
		uintptr(unsafe.Pointer(&cb.buffer[0])) +
			uintptr(idx%cb.capacity)*8,
	)) = latencyMs

	if cb.size.Load() < cb.capacity {
		cb.size.Add(1)
	}
}

// last value written
func (cb *CircularBuffer) ReadLatest() float64 {
	idx := (cb.writeIndex.Load() - 1 + cb.capacity) % cb.capacity
	return cb.buffer[idx]
}

// snapshot current contents
func (cb *CircularBuffer) DrainAll() []float64 {
	count := cb.size.Load()
	if count == 0 {
		return []float64{}
	}
	snapshot := make([]float64, count)
	startIdx := (cb.writeIndex.Load() - count + cb.capacity*2) % cb.capacity
	for i := int64(0); i < count; i++ {
		snapshot[i] = cb.buffer[(startIdx+i)%cb.capacity]
	}
	return snapshot
}

func (cb *CircularBuffer) Size() int64     { return cb.size.Load() }
func (cb *CircularBuffer) Capacity() int64 { return cb.capacity }

File: probe-java/src/main/java/com/distro/probe/CircularBuffer.java
package com.distro.probe;

import java.util.concurrent.atomic.AtomicInteger;

// Pre-allocated ring buffer. Fixed size array claimed at startup,
// never reallocated. Keeps GC out of the hot path.
public class CircularBuffer {

    // claim memory once, never again
    private final double[] buffer;
    private final int capacity;

    // atomic so multiple threads can write without locking
    private final AtomicInteger writeIndex = new AtomicInteger(0);
    private final AtomicInteger size       = new AtomicInteger(0);

    public CircularBuffer(int capacity) {
        this.capacity = capacity;
        this.buffer   = new double[capacity];
    }

    // write next value, wrap around if full
    public void write(double latencyMs) {
        int idx = writeIndex.getAndIncrement() % capacity;
        buffer[idx] = latencyMs;
        if (size.get() < capacity) {
            size.incrementAndGet();
        }
    }

    // latest value written
    public double readLatest() {
        int idx = (writeIndex.get() - 1 + capacity) % capacity;
        return buffer[idx];
    }

    // snapshot everything currently in the buffer
    public double[] drainAll() {
        int count = size.get();
        double[] snapshot = new double[count];
        int startIdx = (writeIndex.get() - count + capacity * 2) % capacity;
        for (int i = 0; i < count; i++) {
            snapshot[i] = buffer[(startIdx + i) % capacity];
        }
        return snapshot;
    }

    public int size()     { return size.get(); }
    public int capacity() { return capacity;   }
}
